internal/mail/gmail/gmailauth.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package gmail
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/smtp"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/zeebo/errs"
+)
+
+// Auth is XOAUTH2 implementation of smtp.Auth interface for gmail
+type Auth struct {
+	UserEmail string
+
+	Storage *TokenStore
+}
+
+// Start returns proto and auth credentials for first auth msg
+func (auth *Auth) Start(server *smtp.ServerInfo) (proto string, toServer []byte, err error) {
+	token, err := auth.Storage.Token()
+	if err != nil {
+		return "", nil, err
+	}
+
+	format := fmt.Sprintf("user=%s\001auth=%s %s\001\001", auth.UserEmail, token.Type, token.AccessToken)
+	return "XOAUTH2", []byte(format), nil
+}
+
+// Next sends empty response to solve SASL challenge if response code is 334
+func (auth *Auth) Next(fromServer []byte, more bool) (toServer []byte, err error) {
+	if more {
+		return make([]byte, 0), nil
+	}
+
+	return nil, nil
+}
+
+// Token represents OAUTH2 token
+type Token struct {
+	AccessToken  string    `json:"access_token"`
+	RefreshToken string    `json:"refresh_token"`
+	Type         string    `json:"token_type"`
+	Expiry       time.Time `json:"expiry"`
+}
+
+// Credentials represents OAUTH2 credentials
+type Credentials struct {
+	ClientID     string `json:"client_id"`
+	ClientSecret string `json:"client_secret"`
+	TokenURI     string `json:"token_uri"`
+}
+
+// TokenStore is a thread safe storage for OAUTH2 token and credentials
+type TokenStore struct {
+	mu    sync.Mutex
+	token Token
+	creds Credentials
+}
+
+// NewTokenStore creates new instance of token storage
+func NewTokenStore(creds Credentials, token Token) *TokenStore {
+	return &TokenStore{
+		token: token,
+		creds: creds,
+	}
+}
+
+// Token retrieves token in a thread safe way and refreshes it if needed
+func (s *TokenStore) Token() (*Token, error) {
+	s.mu.Lock()
```

Comment:
use `defer s.mu.Unlock()`, it ensures that locks get closed even in the presence of panics

Reply:
agree